### PR TITLE
chore: Read Curve factory pools [TKR-316]

### PR DESCRIPTION
**Pull request by dekz**
## Description

Noticed the Curve API (at least in my exploration) is missing factory pools. For example, `BTRFLY` found [here](https://etherscan.io/address/0xF43b15Ab692fDe1F9c24a9FCE700AdCC809D5391) isn't available in the Curve API. Seems like the `YFI` [pool](https://etherscan.io/address/0xc26b89a667578ec7b3f11b2f98d6fd15c07c54ba) isn't available in the CurveAPI either.

[getFactoryPools](https://api.curve.fi/api/getFactoryPools)

Guessing it only works off pools which have been [registered](https://etherscan.io/address/0x90E00ACe148ca3b23Ac1bC8C240C2a7Dd9c2d7f5) via [getRegistryAddress](https://api.curve.fi/api/getRegistryAddress), Or are constructed from some other factory.

[Curve Factory Crypto](https://github.com/curvefi/curve-factory-crypto/blob/e2a59ab163b5b715b38500585a5d1d9c0671eb34/deployment-logs/deployment.log)

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Review by dextracker — COMMENTED** [review]
LGTM!

File: packages/asset-swapper/contracts/src/curve/CurvePoolFactoryReader.sol [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        external
+        view
+        returns (CurveFactoryPool memory poolInfo)
+    {
+        poolInfo.pool = address(pool);
+
+        // All pools seem to have 2 tokens
+        poolInfo.coins = new address[](2);
+        poolInfo.coins[0] = pool.coins(0);
+        poolInfo.coins[1] = pool.coins(1);
```
Comment (shawnxin89) [inline packages/asset-swapper/contracts/src/curve/CurvePoolFactoryReader.sol:82]: I think there are some factory pools (e.g. 3EURpool) with more than 2 tokens. We could probably use `pool.N_COINS` to get the number of tokens in the pool.

Comment (dekz) [reply packages/asset-swapper/contracts/src/curve/CurvePoolFactoryReader.sol]: Improved it by walking through until it reverts/returns `address(0)`. Couldn't access `pool.N_COINS` publicly.

**Review by shawnxin89 — APPROVED** [review]
LGTM 👍

**Comment by elenadimitrova** [comment]
Closing. Do not delete branch.

**Pull request closed without merging**